akka-docs/src/main/paradox/scala/stream/stream-substream.md

```diff
@@ -0,0 +1,125 @@
+# Substreams
+
+Substreams are represented as `SubSource` or `SubFlow` instances, on which you can multiplex a single `Source` or `Flow`
+into a stream of streams.
+
+SubFlows cannot contribute to the super-flow’s materialized value since they are materialized later,
+during the runtime of the flow graph processing.
+
+Stages that creates substreams are listed on @ref:[Nesting and flattening stages](stages-overview.md#nesting-and-flattening-stages)
+
+## Nesting stages
+
+### groupBy
+
+A typical operation that generates substreams is `groupBy`.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy1 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy1 }
+
+![stream-substream-groupBy1.png](../../images/stream-substream-groupBy1.png)
+
+This operation splits the incoming stream into separate output
+streams, one for each element key. The key is computed for each element
+using the given function, which is `f` in the above diagram. When a new key is encountered for the first time
+a new substream is opened and subsequently fed with all elements belonging to that key.
+
+If you add a `Sink` or `Flow` right after the `groupBy` stage,
+all transformations are applied to all encountered substreams in the same fashion.
+So, if you add the following `Sink`, that is added to each of the substreams as in the below diagram.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy2 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy2 }
+
+![stream-substream-groupBy2.png](../../images/stream-substream-groupBy2.png)
+
+Also substreams, more precisely, `SubFlow` and `SubSource` have methods that allow you to
+merge or concat substreams into the master stream again.
+
+The `mergeSubstreams` method merges an unbounded number of substreams back to the master stream.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy3 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy3 }
+
+![stream-substream-groupBy3.png](../../images/stream-substream-groupBy3.png)
+
+You can limit the number of active substreams running and being merged at a time,
+with either the `mergeSubstreamsWithParallelism` or `concatSubstreams` method.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy4 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy4 }
+
+However, since the number of active substreams is capped,
+be careful so that these methods do not cause deadlocks with back pressure like in the below diagram.
```

Comment:
By default (all, I think at least) substreams run in the same actor, so there is only _really_ one stage at a time active, I wonder if we should stress this in the beginning, that to get the sub streams to run in parallell you have to add `async` boundaries after the sub stream-stages are created.

When I read this paragraph it made me think that if I didn't already know this I would infer that active means running-in-paralell.

Reply:
Changed from "active" to "running (i.e. not yet completed)".

By the way,  I got the original description from the Scaladoc comments of [SubFlow](https://github.com/akka/akka/blob/dde564797ee8406bfb6b00ea519d57a6bac4822e/akka-stream/src/main/scala/akka/stream/scaladsl/SubFlow.scala#L29). Should I also change it there?
```
   * Flatten the sub-flows back into the super-flow by performing a merge
   * without parallelism limit (i.e. having an unbounded number of sub-flows
   * active concurrently).
```